src/common/hist_util.cu

```diff
@@ -372,69 +372,74 @@ struct GPUSketcher {
+
     // create device shards
     shards_.resize(dist.Devices().Size());
     dh::ExecuteIndexShards(&shards_, [&](int i, std::unique_ptr<DeviceShard>& shard) {
         size_t start = dist.ShardStart(batch.Size(), i);
         size_t size = dist.ShardSize(batch.Size(), i);
         shard = std::unique_ptr<DeviceShard>(
             new DeviceShard(dist.Devices().DeviceId(i), start,
-                            start + size, param_, sketch_container));
+                            start + size, param_, sketch_container_.get()));
       });
 
     // compute sketches for each shard
     dh::ExecuteIndexShards(&shards_,
                            [&](int idx, std::unique_ptr<DeviceShard>& shard) {
                              shard->Init(batch, info, gpu_batch_nrows_);
                              shard->Sketch(batch, info);
+                             shard->ComputeRowStride();
                            });
-    size_t row_stride = 0;
+
+    // compute row stride across all shards
     for (const auto &shard : shards_) {
-      row_stride = std::max(row_stride, shard->GetRowStride());
+      row_stride_ = std::max(row_stride_, shard->GetRowStride());
     }
-
-    return row_stride;
   }
 
   GPUSketcher(const tree::TrainParam &param, int gpu_nrows)
-    : param_(param), gpu_batch_nrows_(gpu_nrows) {
+    : param_(param), gpu_batch_nrows_(gpu_nrows), row_stride_(0) {
   }
 
-  /* Builds the sketches on the GPU */
+  /* Builds the sketches on the GPU for the dmatrix */
   size_t Sketch(DMatrix *dmat, HistCutMatrix *hmat) {
-    size_t row_stride = 0;
     const MetaInfo &info = dmat->Info();
 
-    // Initialize Sketches for this dmatrix
-    SketchContainer sketch_container;
-    sketch_container.sketches_.resize(info.num_col_);
-    sketch_container.col_locks_.resize(info.num_col_);
-#pragma omp parallel for schedule(static) if (info.num_col_ > 1000)
-    for (int icol = 0; icol < info.num_col_; ++icol) {
-      sketch_container.sketches_[icol].Init(info.num_row_, 1.0 / (8 * param_.max_bin));
-      sketch_container.col_locks_[icol].reset(new std::mutex);
-    }
-
+    row_stride_ = 0;
+    sketch_container_.reset(new SketchContainer(param_, dmat));
     for (const auto &batch : dmat->GetRowBatches()) {
-      GPUDistribution dist =
-        GPUDistribution::Block(GPUSet::All(param_.gpu_id, param_.n_gpus, batch.Size()));
-      size_t batch_row_stride = this->SketchBatch(dist, batch, info, &sketch_container);
-      row_stride = std::max(row_stride, batch_row_stride);
+      this->SketchBatch(batch, info);
     }
 
-    hmat->Init(&sketch_container.sketches_, param_.max_bin);
+    hmat->Init(&sketch_container_.get()->sketches_, param_.max_bin);
 
-    return row_stride;
+    return row_stride_;
   }
 
  private:
   std::vector<std::unique_ptr<DeviceShard>> shards_;
   const tree::TrainParam &param_;
   int gpu_batch_nrows_;
+  size_t row_stride_;
+  std::unique_ptr<SketchContainer> sketch_container_;
 };
 
 size_t DeviceSketch
```

Comment:
Please document what value is being returned.

Alternatively, consider using a pointer parameter instead of returning a value.